<a href="https://colab.research.google.com/github/vhvaldiviesor/Programacion-para-analitica-descriptiva-y-predictiva/blob/main/Actividad_Calificada_07_Datos_Duplicados_269711.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | Victor Hugo Valdivieso Rodriguez |
| **Matrícula** | 45210 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [13]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [14]:
# keep='first': marca como duplicada cada copia EXCEPTO la primera aparición de cada grupo
# -> nos dice cuántas filas "sobran" si conserváramos solo la primera ocurrencia de cada reserva repetida
dup_first = hoteles.duplicated(keep='first').sum()

# keep='last': marca como duplicada cada copia EXCEPTO la última aparición de cada grupo
# -> el mismo número de filas "sobrantes", pero contadas desde el otro extremo del grupo
dup_last = hoteles.duplicated(keep='last').sum()

# keep=False: marca como duplicada TODA fila que pertenece a un grupo con más de una copia
# (incluida la primera y la última) -> nos dice cuántas filas están *involucradas* en algún duplicado
dup_false = hoteles.duplicated(keep=False).sum()

print(f"keep='first' -> {dup_first} filas duplicadas (sobrantes si se conserva la primera copia de cada grupo)")
print(f"keep='last'  -> {dup_last} filas duplicadas (sobrantes si se conserva la última copia de cada grupo)")
print(f"keep=False   -> {dup_false} filas duplicadas (TODAS las copias de cada grupo repetido, sin conservar ninguna)")

keep='first' -> 31994 filas duplicadas (sobrantes si se conserva la primera copia de cada grupo)
keep='last'  -> 31994 filas duplicadas (sobrantes si se conserva la última copia de cada grupo)
keep=False   -> 40165 filas duplicadas (TODAS las copias de cada grupo repetido, sin conservar ninguna)


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [15]:
# Porcentaje de filas sobrantes: usamos dup_first (o dup_last, da el mismo número),
# ya que ambos representan cuántas filas se eliminarían al quedarnos con una sola copia por grupo
pct_sobrantes = dup_first / len(hoteles) * 100
print(f'Porcentaje de filas sobrantes: {pct_sobrantes:.2f}%')

# Número de grupos repetidos: la diferencia entre "todas las copias" (keep=False) y
# "las copias sobrantes" (keep='first') es exactamente el número de grupos, porque cada
# grupo de tamaño n aporta n filas a keep=False y (n-1) filas a keep='first' -- la resta
# deja 1 fila por grupo, es decir, el número de grupos.
n_grupos_repetidos = dup_false - dup_first
print(f'Número de grupos repetidos: {n_grupos_repetidos}')

# Verificación independiente con value_counts(), para confirmar el cálculo anterior
conteos_combinaciones = hoteles.value_counts(dropna=False)
n_grupos_verificacion = (conteos_combinaciones > 1).sum()
print(f'Verificación con value_counts(): {n_grupos_verificacion} grupos repetidos')

Porcentaje de filas sobrantes: 26.80%
Número de grupos repetidos: 8171
Verificación con value_counts(): 8171 grupos repetidos


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [16]:
hoteles_copia = hoteles.copy()
# Marcamos cada fila como duplicada (sobrante) o no, usando keep='first' igual que en 1.1
hoteles_copia['es_duplicado'] = hoteles_copia.duplicated(keep='first')

# Agrupamos por hotel: sum() cuenta cuántos True (duplicados) hay, count() da el total de reservas
duplicados_por_hotel = hoteles_copia.groupby('hotel')['es_duplicado'].agg(
    filas_duplicadas='sum',
    total_reservas='count'
)
duplicados_por_hotel['pct_duplicadas'] = (
    duplicados_por_hotel['filas_duplicadas'] / duplicados_por_hotel['total_reservas'] * 100
)

duplicados_por_hotel

,filas_duplicadas,total_reservas,pct_duplicadas
hotel,,,
City Hotel,25902,79330,32.650952
Resort Hotel,6092,40060,15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

El problema de duplicados en este dataset es grave: 31,994 de las 119,390 filas (26.80%) son copias sobrantes de otra fila idéntica en las 32 columnas, agrupadas en 8,171 grupos repetidos. Es decir, más de una de cada cuatro reservas del dataset es una copia de otra reserva que ya está registrada, lo cual es una proporción demasiado alta para ignorarla en cualquier análisis o modelo.

El problema tampoco afecta por igual a ambos hoteles: City Hotel tiene 25,902 filas duplicadas de 79,330 reservas totales (32.65%), mientras que Resort Hotel tiene 6,092 duplicadas de 40,060 (15.21%). Es decir, en City Hotel casi 1 de cada 3 reservas es un duplicado, más del doble de la proporción que se observa en Resort Hotel. Esto sugiere que la causa de los duplicados (posiblemente reservas grupales o un proceso de captura distinto) se presenta con más frecuencia en el hotel urbano que en el hotel vacacional.

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [17]:
columnas_a_mostrar = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                       'arrival_date_day_of_month', 'adults', 'country', 'market_segment',
                       'deposit_type', 'adr', 'reservation_status']

filas_duplicadas = hoteles[hoteles.duplicated(keep=False)].copy()

# Ordenamos por TODAS las columnas mostradas: como las filas de un mismo grupo son idénticas
# en todas ellas, ordenar por esa combinación garantiza que las copias de cada grupo queden
# juntas en el resultado (a diferencia de ordenar solo por una o dos columnas, donde grupos
# distintos con, por ejemplo, el mismo hotel y lead_time podrían intercalarse).
filas_ordenadas = filas_duplicadas.sort_values(by=columnas_a_mostrar)

filas_ordenadas[columnas_a_mostrar].head(12)

,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40643,City Hotel,0,2015,August,4,2,PRT,Direct,No Deposit,0.0,Canceled
40644,City Hotel,0,2015,August,4,2,PRT,Direct,No Deposit,0.0,Canceled
40772,City Hotel,0,2015,August,7,2,PRT,Direct,No Deposit,75.0,Check-Out
40802,City Hotel,0,2015,August,7,2,PRT,Direct,No Deposit,75.0,Check-Out
40821,City Hotel,0,2015,August,8,2,PRT,Online TA,No Deposit,89.0,Check-Out
40838,City Hotel,0,2015,August,8,2,PRT,Online TA,No Deposit,89.0,Check-Out
76792,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
76793,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
76794,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
41067,City Hotel,0,2015,August,11,1,PRT,Corporate,No Deposit,88.0,Check-Out


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [18]:
# dropna=False es importante: por defecto value_counts() ignora combinaciones con NaN en
# alguna columna, y aquí agent/company tienen muchos faltantes -- sin dropna=False perderíamos
# de vista justamente los grupos duplicados que comparten un valor faltante en esas columnas.
conteos_combinaciones = hoteles.value_counts(dropna=False)

# Nos quedamos solo con las combinaciones que se repiten (tamaño de grupo > 1)
grupos_repetidos = conteos_combinaciones[conteos_combinaciones > 1]

# 1. Cuántos grupos hay de cada tamaño (2 copias, 3 copias, ...) -- top 10 tamaños más frecuentes
print('Cantidad de grupos por tamaño (10 tamaños más frecuentes):')
print(grupos_repetidos.value_counts().head(10))
print()

# 2. Tamaño del grupo más grande y su combinación de valores
tamano_max = grupos_repetidos.max()
combinacion_mas_grande = grupos_repetidos.idxmax()
print(f'El grupo más grande tiene {tamano_max} copias idénticas.')
print('Su combinación de valores es:')
print(combinacion_mas_grande)

Cantidad de grupos por tamaño (10 tamaños más frecuentes):
count
2     5034
3      969
4      497
5      275
6      171
10     123
7      115
8      103
20      97
9       93
Name: count, dtype: int64

El grupo más grande tiene 180 copias idénticas.
Su combinación de valores es:
('City Hotel', np.int64(1), np.int64(277), np.int64(2016), 'November', np.int64(46), np.int64(7), np.int64(1), np.int64(2), np.int64(2), np.float64(0.0), np.int64(0), 'BB', 'PRT', 'Groups', 'TA/TO', np.int64(0), np.int64(0), np.int64(0), 'A', 'A', np.int64(0), 'Non Refund', np.float64(nan), np.float64(nan), np.int64(0), 'Transient', np.float64(100.0), np.int64(0), np.int64(0), 'Canceled', '2016-04-04')


### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [19]:
# Reutilizamos filas_duplicadas (keep=False) de la Actividad 2.1

print('--- market_segment: dataset completo vs. filas duplicadas ---')
print('Completo:')
print((hoteles['market_segment'].value_counts(normalize=True) * 100).round(2))
print()
print('Duplicadas:')
print((filas_duplicadas['market_segment'].value_counts(normalize=True) * 100).round(2))
print()

print('--- deposit_type: dataset completo vs. filas duplicadas ---')
print('Completo:')
print((hoteles['deposit_type'].value_counts(normalize=True) * 100).round(2))
print()
print('Duplicadas:')
print((filas_duplicadas['deposit_type'].value_counts(normalize=True) * 100).round(2))
print()

print('--- Tasa de cancelación: dataset completo vs. filas duplicadas ---')
print(f"Completo:    {hoteles['is_canceled'].mean() * 100:.2f}%")
print(f"Duplicadas:  {filas_duplicadas['is_canceled'].mean() * 100:.2f}%")

--- market_segment: dataset completo vs. filas duplicadas ---
Completo:
market_segment
Online TA        47.30
Offline TA/TO    20.29
Groups           16.59
Direct           10.56
Corporate         4.44
Complementary     0.62
Aviation          0.20
Undefined         0.00
Name: proportion, dtype: float64

Duplicadas:
market_segment
Groups           41.61
Offline TA/TO    30.40
Online TA        20.59
Corporate         3.67
Direct            3.50
Complementary     0.19
Aviation          0.05
Name: proportion, dtype: float64

--- deposit_type: dataset completo vs. filas duplicadas ---
Completo:
deposit_type
No Deposit    87.65
Non Refund    12.22
Refundable     0.14
Name: proportion, dtype: float64

Duplicadas:
deposit_type
No Deposit    64.17
Non Refund    35.64
Refundable     0.19
Name: proportion, dtype: float64

--- Tasa de cancelación: dataset completo vs. filas duplicadas ---
Completo:    37.04%
Duplicadas:  58.39%


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

**a)** La evidencia apunta a que la mayoría de las filas duplicadas son reservas genuinamente distintas que coinciden en todos sus valores registrados, no simples errores de captura -- y esto tiene una explicación de negocio razonable: reservas grupales. Tres hallazgos lo respaldan. Primero, el grupo más grande tiene 180 copias idénticas, todas con `market_segment='Groups'`, `deposit_type='Non Refund'` y `is_canceled=1` -- es perfectamente plausible que un grupo de turistas reserve 180 habitaciones idénticas (mismo tipo de habitación, mismas fechas, mismo país, mismo canal) y que, al no existir un identificador de reserva ni de huésped, cada habitación quede registrada como una fila indistinguible de las demás. Segundo, `market_segment='Groups'` está sobrerrepresentado entre las filas duplicadas (41.61%) frente a su proporción en el dataset completo (16.59%), y lo mismo ocurre con `deposit_type='Non Refund'` (35.64% en duplicadas vs. 12.22% en el total) -- ambos patrones son característicos de reservas grupales gestionadas por agencias, que suelen pagarse con depósito no reembolsable. Tercero, la tasa de cancelación entre las filas duplicadas es de 58.39%, muy por encima del 37.04% del dataset completo -- consistente con que, cuando se cancela una reserva grupal, se cancela el grupo completo de filas idénticas a la vez.

Dicho esto, no se puede descartar que una fracción de los duplicados sí sean errores de captura (por ejemplo, un doble envío del mismo formulario) -- simplemente no hay forma de distinguir ambos casos sin un identificador de reserva, que es justamente la limitación que señala el enunciado.

**b)** `duplicated()` trata dos valores `NaN` en la misma columna como **iguales** entre sí para efectos de comparación (a diferencia de la igualdad estándar de Python/NumPy, donde `NaN != NaN`). Esto tiene una implicación importante para este análisis: la columna `company` tiene 112,593 valores faltantes de 119,390 filas (94.3% de faltantes), así que dos reservas que no tengan ninguna relación entre sí pero que ambas carezcan de un valor en `company` (y coincidan en el resto de las columnas) se considerarán "iguales" en esa columna, aumentando la probabilidad de que sean marcadas como duplicadas. Esto significa que el conteo de duplicados que hemos calculado podría estar sobreestimando ligeramente los duplicados "reales", ya que parte de la coincidencia depende de que ambas filas compartan la ausencia de dato en `company` (y, en menor medida, en `agent`, con 16,340 faltantes), no necesariamente de que se trate de la misma reserva repetida.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [20]:
# keep='first' conserva la primera aparición de cada grupo (igual criterio que usamos en 1.1-1.3);
# reset_index(drop=True) renumera el índice de 0 a n-1, sin dejar huecos por las filas eliminadas
hoteles_limpio = hoteles.drop_duplicates(keep='first').reset_index(drop=True)

filas_antes = len(hoteles)
filas_despues = len(hoteles_limpio)
filas_eliminadas = filas_antes - filas_despues

print(f'Filas antes:      {filas_antes}')
print(f'Filas después:    {filas_despues}')
print(f'Filas eliminadas: {filas_eliminadas}')

Filas antes:      119390
Filas después:    87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [21]:
# Calculamos las medias de is_canceled y adr por hotel, y por separado el número de reservas con .size()
resumen_original = hoteles.groupby('hotel')[['is_canceled', 'adr']].mean().reset_index()
resumen_original['reservas'] = hoteles.groupby('hotel').size().values

resumen_limpio = hoteles_limpio.groupby('hotel')[['is_canceled', 'adr']].mean().reset_index()
resumen_limpio['reservas'] = hoteles_limpio.groupby('hotel').size().values

print('Resumen original (con duplicados):')
print(resumen_original)
print()
print('Resumen limpio (sin duplicados):')
print(resumen_limpio)

Resumen original (con duplicados):
          hotel  is_canceled         adr  reservas
0    City Hotel     0.417270  105.304465     79330
1  Resort Hotel     0.277634   94.952930     40060

Resumen limpio (sin duplicados):
          hotel  is_canceled         adr  reservas
0    City Hotel     0.300386  110.985944     53428
1  Resort Hotel     0.234809   99.025346     33968


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [22]:
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

# pd.concat() apila ambos resúmenes uno debajo del otro (mismo esquema de columnas)
comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True)

# Ordenamos por hotel y version para que las dos versiones de cada hotel queden juntas,
# y renumeramos el índice tras el ordenamiento
comparacion = comparacion.sort_values(by=['hotel', 'version']).reset_index(drop=True)

comparacion

,hotel,is_canceled,adr,reservas,version
0,City Hotel,0.417270,105.304465,79330,original
1,City Hotel,0.300386,110.985944,53428,sin duplicados
2,Resort Hotel,0.277634,94.952930,40060,original
3,Resort Hotel,0.234809,99.025346,33968,sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

**a)** La métrica que más cambia es la tasa de cancelación (`is_canceled`), y el cambio es notoriamente mayor en City Hotel. En City Hotel, la tasa de cancelación baja de 41.73% (original) a 30.04% (sin duplicados), una caída de casi 11.7 puntos porcentuales; en Resort Hotel baja de 27.76% a 23.48%, una caída de solo 4.3 puntos. La tarifa diaria promedio (`adr`) también sube en ambos hoteles al quitar los duplicados (City: de 105.30 a 110.99; Resort: de 94.95 a 99.03), pero el cambio es mucho más modesto que el de la tasa de cancelación.

Esto se explica directamente con lo encontrado en la Actividad 2: ahí vimos que las filas duplicadas tienen una tasa de cancelación mucho más alta que el dataset completo (58.39% vs. 37.04%), consistente con que gran parte de los duplicados son reservas grupales que se cancelan en bloque. Y en la Actividad 1 vimos que City Hotel concentra más del doble de la proporción de duplicados que Resort Hotel (32.65% vs. 15.21%). Al combinar ambos hallazgos: City Hotel tiene muchas más filas "infladas" por reservas grupales canceladas, así que al eliminarlas su tasa de cancelación baja mucho más que la de Resort Hotel, donde el problema de duplicados es menor desde un inicio.

**b)** Entrenar un modelo con los datos originales (sin quitar duplicados) causaría al menos dos problemas serios. Primero, en el entrenamiento: si una fila duplicada (por ejemplo, una de las 180 copias del grupo más grande) termina repartida entre el conjunto de entrenamiento y el conjunto de prueba (algo que ocurre fácilmente con una división aleatoria), el modelo estaría literalmente memorizando en entrenamiento un ejemplo que luego vuelve a ver, sin cambios, en la evaluación -- esto es una forma de fuga de datos (*data leakage*) que infla artificialmente las métricas de desempeño (precisión, recall, etc.), haciendo parecer que el modelo generaliza mejor de lo que realmente generaliza. Segundo, incluso si todos los duplicados de un grupo cayeran del mismo lado de la división, el modelo daría un peso desproporcionado a esos patrones sobrerrepresentados (como las reservas grupales canceladas) simplemente porque aparecen muchas más veces en los datos, sesgando las predicciones hacia ese tipo de reserva en detrimento de patrones menos repetidos pero igualmente válidos.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [23]:
# Debe dar 0: si drop_duplicates() funcionó bien, ya no debe quedar ninguna fila duplicada
print('Duplicados exactos restantes:', hoteles_limpio.duplicated().sum())

# Debe dar 0: reset_index(drop=True) no debería haber dejado índices repetidos
print('Índices repetidos:', hoteles_limpio.index.duplicated().sum())

# Verificamos que el índice sea exactamente el rango esperado, de 0 a len-1, sin huecos
indice_correcto = list(hoteles_limpio.index) == list(range(len(hoteles_limpio)))
print('¿El índice va de 0 a len(hoteles_limpio)-1, sin huecos?:', indice_correcto)

Duplicados exactos restantes: 0
Índices repetidos: 0
¿El índice va de 0 a len(hoteles_limpio)-1, sin huecos?: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [24]:
# Definimos el subset de columnas a comparar: todas menos reservation_status_date, usando
# el parámetro subset de duplicated() para ignorar esa columna en la comparación
columnas_sin_fecha = [c for c in hoteles_limpio.columns if c != 'reservation_status_date']

# 1. Filas involucradas (keep=False) y filas sobrantes (keep='first', mismo criterio que antes)
mascara_parcial_todas = hoteles_limpio.duplicated(subset=columnas_sin_fecha, keep=False)
mascara_parcial_sobrantes = hoteles_limpio.duplicated(subset=columnas_sin_fecha, keep='first')
print(f'Filas involucradas en duplicados parciales: {mascara_parcial_todas.sum()}')
print(f'Filas sobrantes (keep=first):                {mascara_parcial_sobrantes.sum()}')
print()

# 2. Mostramos algunas filas, ordenadas por las columnas comparadas para que las copias queden juntas
filas_parcial = hoteles_limpio[mascara_parcial_todas].copy()
columnas_mostrar_parcial = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                             'arrival_date_day_of_month', 'adr', 'reservation_status',
                             'reservation_status_date']
filas_parcial_ordenadas = filas_parcial.sort_values(by=columnas_sin_fecha)
print('Ejemplos de filas con duplicado parcial (difieren solo en la fecha de estado):')
print(filas_parcial_ordenadas[columnas_mostrar_parcial].head(12).to_string())
print()

# 3. Distribución de reservation_status en esas filas
print('Distribución de reservation_status en las filas con duplicado parcial:')
print(filas_parcial['reservation_status'].value_counts())

Filas involucradas en duplicados parciales: 471
Filas sobrantes (keep=first):                245

Ejemplos de filas con duplicado parcial (difieren solo en la fecha de estado):
            hotel  lead_time  arrival_date_year arrival_date_month  arrival_date_day_of_month     adr reservation_status reservation_status_date
35572  City Hotel          9               2015          September                         10   95.00           Canceled              2015-09-03
35574  City Hotel          9               2015          September                         10   95.00           Canceled              2015-09-01
51267  City Hotel         12               2017               July                         15  139.00           Canceled              2017-07-10
51268  City Hotel         12               2017               July                         15  139.00           Canceled              2017-07-07
54170  City Hotel         14               2015            October                         15  100

**Respuesta 4.2:**

Estas 471 filas (agrupadas en 245 sobrantes) muy probablemente representan la misma reserva registrada más de una vez con una fecha de cambio de estado distinta, no reservas nuevas: el 100% de estas filas tiene `reservation_status='Canceled'` (0% en cualquier otro estado), y en los ejemplos mostrados se ve el mismo patrón repetidamente -- filas idénticas en hotel, fechas de llegada, tarifa y todo lo demás, salvo por `reservation_status_date`, que cambia de un día a otro cercano (por ejemplo, dos filas canceladas para el mismo `lead_time`, mismo `arrival_date`, pero con fechas de cancelación registradas el 2015-09-01 y el 2015-09-03). Esto es consistente con que el proceso que generó los datos haya tomado "fotografías" (snapshots) del estado de la reserva en distintos momentos, o con que el sistema haya vuelto a registrar la cancelación al reprocesar o confirmar el estado en una fecha posterior, sin que eso implique una reserva real adicional.

**¿Las eliminaría?** Sí, las trataría como duplicados y me quedaría con una sola fila por combinación -- pero, a diferencia del criterio de la Actividad 3 (que conserva arbitrariamente la primera fila), aquí sí importa la fecha: me quedaría con la fila que tenga la `reservation_status_date` **más reciente**, ya que probablemente representa el registro final o más actualizado del estado de esa reserva, en vez de tomar la primera aparición sin más. Antes de aplicar esa limpieza en un proyecto real, documentaría esta decisión explícitamente (como se pide en 4.3), porque a diferencia de los duplicados exactos de la Actividad 1, aquí sí estamos ignorando una columna a propósito, y otro analista podría razonablemente preferir quedarse con la fecha más antigua (el primer registro del evento) en vez de la más reciente.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

**Registro de limpieza de duplicados — dataset Hotel Booking Demand**

**Criterio:** se consideró duplicada toda fila que coincidiera con otra en las 32 columnas del dataset (comparación exacta, sin excluir ninguna columna). No existe un identificador de reserva en el dataset original, por lo que este criterio de "coincidencia total" fue el único disponible para detectar copias.

**Regla de conservación:** se usó `keep='first'`, conservando la primera fila de cada grupo repetido y descartando el resto. En este caso la elección de `keep='first'` frente a `keep='last'` es prácticamente irrelevante para el resultado agregado, porque las filas de un mismo grupo son idénticas en todas las columnas (incluyendo `reservation_status`, `adr`, etc.) -- conservar la primera o la última no cambia ninguna métrica descriptiva calculada sobre `hoteles_limpio`, ya que se conserva exactamente el mismo valor sin importar cuál copia se elija. (Esto es distinto al caso de los duplicados parciales de la Actividad 4.2, donde sí importa cuál fila se conserva, porque difieren en `reservation_status_date`.)

**Resultado:** el dataset original tenía 119,390 filas. Tras eliminar los 31,994 duplicados exactos (26.80% del total, agrupados en 8,171 grupos repetidos), `hoteles_limpio` quedó con 87,396 filas.

**Supuestos y limitaciones:**
1. Se asumió que una coincidencia exacta en las 32 columnas es suficiente evidencia de que dos filas son "la misma información repetida", aun sin poder confirmarlo con un identificador único -- como se discutió en 2.4, es más plausible que se trate de reservas grupales genuinas que de errores de captura, pero no se puede descartar del todo esta segunda posibilidad.
2. `duplicated()` trata dos valores `NaN` como iguales; dado que `company` tiene 94.3% de faltantes, es posible que el conteo de duplicados esté ligeramente sobreestimado por filas que solo coinciden en la ausencia de ese dato.
3. Este proceso limpia únicamente duplicados exactos; no se aplicó en esta etapa el criterio de duplicado parcial (ignorando `reservation_status_date`) que se exploró en la Actividad 4.2 -- de aplicarse en un flujo de limpieza completo, debería documentarse como un paso adicional y explícito, con su propia justificación.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

**1.** No convendría eliminar los duplicados exactos si el objetivo del análisis fuera, por ejemplo, estimar cuántas *habitaciones* se reservaron en total (en vez de cuántas *reservas únicas* hubo). Como vimos en la Actividad 2, el grupo más grande de duplicados corresponde muy probablemente a una reserva grupal de 180 habitaciones idénticas -- si el análisis busca calcular ocupación total, ingresos totales o demanda de habitaciones, eliminar esas 180 filas y quedarnos con 1 subestimaría gravemente la cifra real, porque cada fila probablemente representa una habitación distinta dentro del mismo grupo, no la misma habitación registrada varias veces.

**2.** Pediría, sobre todo, un identificador único de reserva (`booking_id`) y, si es posible, un identificador de huésped o de grupo (`group_id`). Con un identificador de reserva podríamos confirmar de forma definitiva si dos filas idénticas son la misma reserva capturada dos veces (verdadero duplicado, a eliminar) o dos reservas distintas que solo coinciden en sus atributos (como el caso de las 180 habitaciones grupales, que no deberían eliminarse). Esto resolvería la principal limitación señalada en el registro de decisión (4.3): actualmente estamos infiriendo la naturaleza de los duplicados de forma indirecta (por patrones en `market_segment`, `deposit_type` y tasa de cancelación), cuando un identificador único lo confirmaría con certeza.

**3.** Que `duplicated().sum()` dé 0 solo garantiza que no hay filas que coincidan en **todas** las columnas consideradas -- no garantiza que el dataset esté libre de registros redundantes o inconsistentes en un sentido más amplio. Esto lo confirmamos directamente en la Actividad 4.2: después de limpiar `hoteles_limpio` (con `duplicated().sum() == 0`), todavía encontramos 471 filas que eran esencialmente la misma reserva registrada dos veces, solo que con una columna (`reservation_status_date`) ligeramente distinta -- bastaba con que una sola columna cambiara, por mínimo que fuera el cambio, para que `duplicated()` (con todas las columnas) dejara de detectarlas como copias. En otras palabras, `duplicated()` por defecto es tan estricto como las columnas que se le pasen: si el verdadero duplicado tiene alguna variación en una columna (una fecha distinta, un espacio de más, un identificador temporal), un conteo de "0 duplicados exactos" puede dar una falsa sensación de que los datos están limpios, cuando en realidad haría falta revisar duplicados *parciales* (con `subset`) o aplicar normalización previa para detectarlos.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.